File: utils/src/main/java/com/lab/azure/bootloader/utils/PlanHelper.java
package com.lab.azure.bootloader.utils;

import com.lab.azure.bootloader.utils.arguments.PlanArgument;
import com.microsoft.azure.storage.CloudStorageAccount;
import com.microsoft.azure.storage.StorageException;
import com.microsoft.azure.storage.table.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.net.URISyntaxException;
import java.security.InvalidKeyException;
import java.util.Date;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

public class PlanHelper {
    public static void makePlan(PlanArgument planArgument) throws IOException, URISyntaxException, InvalidKeyException, StorageException {
        Logger logger = LoggerFactory.getLogger("plan");
        logger.info("plan {} start", planArgument.getTask());
        Settings settings = Settings.loadCfg(planArgument.getSettingFile());
        if (!settings.validate()) {
            logger.warn("settings error");
            System.exit(-1);
        }
        if (!checkKey(planArgument.getTask())) {
            logger.warn("task key {} error", planArgument.getTask());
            System.exit(-1);
        }
        CloudStorageAccount account = CloudStorageAccount.parse(settings.getAzureWebJobsStorage());
        CloudTableClient tableClient = account.createCloudTableClient();
        CloudTable jobTable = tableClient.getTableReference(settings.getJobTable());
        jobTable.createIfNotExists();

        {
            String filter;
            {
                String pkFilter = TableQuery.generateFilterCondition("PartitionKey", TableQuery.QueryComparisons.EQUAL, planArgument.getTask());
                String rkFilter = TableQuery.generateFilterCondition("RowKey", TableQuery.QueryComparisons.EQUAL, planArgument.getTask());
                filter = TableQuery.combineFilters(pkFilter, TableQuery.Operators.AND, rkFilter);
            }

            TableQuery<JobEntity> query = TableQuery.from(JobEntity.class)
                .where(filter);

            Iterable<JobEntity> entities = jobTable.execute(query);

            List<JobEntity> tasks = StreamSupport.stream(entities.spliterator(), false).collect(Collectors.toList());
            if (!tasks.isEmpty()) {
                logger.warn("task {} exist", planArgument.getTask());
                System.exit(-1);
            }
            if (planArgument.getCount() <= 0) {
                logger.warn("task count {} error", planArgument.getCount());
                System.exit(-1);
            }
        }

        {
            JobEntity jobEntity = new JobEntity();
            jobEntity.setPartitionKey(planArgument.getTask());
            jobEntity.setRowKey(planArgument.getTask());
            jobEntity.setVersion(planArgument.getVersion());
            jobEntity.setAmount(planArgument.getCount());
            jobEntity.setCurrent(0);
            jobEntity.updateTimestamp();
            TableOperation operation = TableOperation.insert(jobEntity);
            jobTable.execute(operation);

        }
        logger.info("plan {} finish", planArgument.getTask());
    }

    private static boolean checkKey(String keyString) {
        try {
            UUID uuid = UUID.fromString(keyString);
            return true;
        } catch (IllegalArgumentException e) {
            return false;
        }
    }
}

File: src/AzureBootloaderCompiler.Utils/PlanHelper.cs
using Microsoft.VisualBasic.CompilerServices;
using Microsoft.WindowsAzure.Storage;
using Microsoft.WindowsAzure.Storage.Table;
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;
using AzureBootloaderCompiler.Utils.Arguments;

namespace AzureBootloaderCompiler.Utils
{
    class PlanHelper
    {
        public static async Task MakePlanAsync(PlanArgument planArgument) 
        {
            Console.WriteLine($"plan {planArgument.Task} start");
            Settings settings = Settings.LoadCfg(planArgument.SettingFile);

            if (!CheckKey(planArgument.Task))
            {
                Console.WriteLine($"task key {planArgument.Task} error");
                Environment.Exit(-1);
            }

            var account = CloudStorageAccount.Parse(settings.AzureWebJobsStorage);
            var tableClient = account.CreateCloudTableClient();
            var cloudTable = tableClient.GetTableReference(settings.JobTable);
            _ = cloudTable.CreateIfNotExistsAsync();
            Console.WriteLine($"Table {settings.JobTable} create success");

            {
                var partitionKey = planArgument.Task;
                var rowKey = planArgument.Task;
                TableOperation retrieveOperation = TableOperation.Retrieve<JobEntity>(partitionKey, rowKey);
                TableResult result = await cloudTable.ExecuteAsync(retrieveOperation);
                
                if (result.Result is JobEntity)
                {
                    Console.WriteLine($"task {planArgument.Task} exist");
                    Environment.Exit(-1);
                }

                
                if (planArgument.Count <= 0)
                {
                    Console.WriteLine($"task count {planArgument.Count} error");
                    Environment.Exit(-1);
                }
            }

            {
                JobEntity jobEntity = new JobEntity();
                jobEntity.PartitionKey = planArgument.Task;
                jobEntity.RowKey = planArgument.Task;
                jobEntity.Version = planArgument.Version;
                jobEntity.Amount = planArgument.Count;
                jobEntity.Current = 0;
                jobEntity.CreateAt = DateTime.Now;
                jobEntity.UpdateAt = DateTime.Now;

                TableOperation insertOrMergeOperation = TableOperation.InsertOrMerge(jobEntity);
                await cloudTable.ExecuteAsync(insertOrMergeOperation);
                Console.WriteLine($"plan {planArgument.Task} stop");
            }
        }

        private static Boolean CheckKey(string keyString)
        {
            try
            {
                Guid uuid = Guid.Parse(keyString);
                return true;
            }
            catch (FormatException)
            {
                return false;
            }
        }
    }
}
